[SPA] Proxy cleanups

javiercn opened the pull request:
There were a couple of gaps in the new SPA experience when running from VS that this PR addresses.
1) Running without debugging (CTRL+F5). Visual Studio would fail to wait until the app was up and running.
2) Switching to IIS Express and launching from Visual Studio would have a similar effect.

This PR introduced a slightly modified approach to better accommodate for these scenarios:
* It reverts back to launching the app in the backend URL directly.
* It automatically detects the backend urls on the proxy scripts so its possible to switch between IIS Express and Kestrel without having to manually update the proxy configuration.
* It introduces a "Startup" page that we display while the proxy is getting ready and that auto-refreshes every 3 seconds.
  * Once the proxy is ready or we've given up, it redirects to the proxy server URL.


https://user-images.githubusercontent.com/6995051/119511753-93719200-bd27-11eb-8a58-44ffa2249838.mp4

javiercn committed:
Add description to package

javiercn committed:
Automatically resolve proxy target based on available environment variables

javiercn committed:
Introduce a temporary page while the proxy is launching and redirect to the proxy page when visiting the backend for the first time

javiercn commented on src/Middleware/Spa/SpaProxy/src/ConfigureSpaDevelopmentServerOptions.cs:
```diff
@@ -0,0 +1,21 @@
+                .Build();
+            configuration.GetSection("SpaProxyServer").Bind(options);
+        }
+    }
```
This was inside `SpaProxyLaunchManager`. 

I've separated it into its own class as I've started using the `IOptions` pattern for the `SpaDevelopmentServerOptions.`

HaoK commented on src/Middleware/Spa/SpaProxy/src/SpaProxyMiddleware.cs:
```diff
@@ -0,0 +1,90 @@
+</head>
+<body>
+  <h1>Launching the SPA proxy...</h1>
+  <p>This page will automatically refresh in one second.</p>
```
Do we have to worry about any localization issues?

javiercn replied:
I don't think so, we don't localize other similar things like the error page and so on.

We can replace this with an icon or some SVG + some CSS, but I don't think is necessary.

HaoK commented on src/Middleware/Spa/SpaProxy/src/SpaHostingStartup.cs:
```diff
@@ -19,6 +21,9 @@ public void Configure(IWebHostBuilder builder)
                 {
                     services.AddHostedService<SpaProxyLaunchManager>();
+                    services.AddSingleton<SpaProxyStatus>();
+                    services.TryAddEnumerable(ServiceDescriptor.Singleton<IConfigureOptions<SpaDevelopmentServerOptions>, ConfigureSpaDevelopmentServerOptions>());
```
Can consider just doing the IConfigure inline here to avoid the class:

something like this would let you avoid the internal class and simplify the bind using the configuration overload which binds implicitly
```
            var configuration = new ConfigurationBuilder()
                .AddJsonFile(Path.Combine(AppContext.BaseDirectory, "spa.proxy.json"))
                .Build();
            services.Configure<SpaDevelopmentServerOptions>(configuration.GetSection("SpaProxyServer");
```

javiercn committed:
* Switch to probing the proxy on a per-request basis. This helps in cases
where the user closes the proxy window (for example when the app is run
without debugging via Ctrl+F5) in IIS Express scenarios.

Addressed feedback

Tratcher commented on src/Middleware/Spa/SpaProxy/src/SpaProxyMiddleware.cs:
```diff
@@ -0,0 +1,92 @@
+</head>
+<body>
+  <h1>Launching the SPA proxy...</h1>
+  <p>This page will automatically redirect to <a href=""{options.ServerUrl}"">{options.ServerUrl}</a> when the SPA proxy is ready.</p>
```
escaping?

The text implies the navigation is automatic, why give a link? It's probably better to remove the href and only show the url text.

javiercn replied:
The navigation is automatic (the client polls with the `http-equiv` meta tag every three seconds). You can click the link to go to the SPA front-end directly if you don't want to wait. We display this page to keep VS happy while launching the SPA CLI proxy.

Tratcher replied:
Then you'll at least need to properly escape the value.

javiercn replied:
@Tratcher I already did, are you referring to something else?

BTW, this will just hold `https://localhost:5002`

Tratcher commented on src/Middleware/Spa/SpaProxy/src/SpaProxyStartupFilter.cs:
```diff
@@ -0,0 +1,28 @@
+            return builder =>
+            {
+                var env = builder.ApplicationServices.GetRequiredService<IWebHostEnvironment>();
+                if (env.IsDevelopment())
```
We need to stop using IsDevelopment in framework libraries, it should only be used in application code since they're the ones that define the environments. This injection logic should be controlled by user settable options instead.
https://github.com/dotnet/aspnetcore/issues/28937

javiercn replied:
This check is here as an additional safeguard to ensure this never gets used outside of development. There is already another check that happens to prevent this injection in the first place, this was done as an additional check.

In general the issue here is that we don't have a great way to determine when you are in a `development environment` (aka running with `dotnet run`) over some other environment.

Things like this package, or the runtime support for Static Web Assets are specifically tailored to enable development scenarios and we want to go as far as possible to avoid involving them outside of those contexts.

javiercn replied:
I've removed the check since its an additional guarantee, that said, I've left some comments on the issue, we keep "complaining" about this without offering a solution.

Tratcher replied:
> There is already another check that happens to prevent this injection in the first place, this was done as an additional check.

What's the other check?

The proposal for 6.0 is to change the framework usage of IsDevelopment so there is user targeted option for each feature and we fall back to IsDevelopment if the option is not set. How you specify that option does depend on the feature setup.

Tratcher replied:
What's the workaround for someone that does rename their dev environment?
`builder.UseMiddleware<SpaProxyMiddleware>();` is internal so they can't add it manually.

javiercn replied:
All the services only get setup if there's a file in the app running folder, which gets generated at build time and not copied to the publish folder.

https://github.com/dotnet/aspnetcore/pull/32997/files#diff-1c1e51e26ffaf3976061683dbfd13701a7184b1d8d1191169aeda59b3003a97aR22

javiercn replied:
The way this is plugged in is via a HostingStartup assembly in launchSettings.json and a file we generate during build. You don't opt-in to individual pieces. It is just meant to launch the CLI automatically and redirect the user to the proxy url (the ServerUrl, which is always localhost:XXX).

There's no public API on purpose since this is just a convenience to keep parity with the old SPA template experience.

javiercn committed:
Address feedback

Tratcher commented on src/Middleware/Spa/SpaProxy/src/SpaProxyMiddleware.cs:
```diff
@@ -0,0 +1,93 @@
+</head>
+<body>
+  <h1>Launching the SPA proxy...</h1>
+  <p>This page will automatically redirect to <a href=""{options.ServerUrl}"">{HtmlEncoder.Default.Encode(options.ServerUrl)}</a> when the SPA proxy is ready.</p>
```
Don't you need to encode both usages of ServerUrl?

javiercn replied:
Done

javiercn commented on src/Middleware/Spa/SpaProxy/src/SpaProxyMiddleware.cs:
```diff
@@ -0,0 +1,93 @@
+</head>
+<body>
+  <h1>Launching the SPA proxy...</h1>
+  <p>This page will automatically redirect to <a href=""{options.ServerUrl}"">{HtmlEncoder.Default.Encode(options.ServerUrl)}</a> when the SPA proxy is ready.</p>
```
```suggestion
  <p>This page will automatically redirect to <a href=""{HtmlEncoder.Default.Encode(options.ServerUrl)}"">{HtmlEncoder.Default.Encode(options.ServerUrl)}</a> when the SPA proxy is ready.</p>
```

javiercn committed:
Update src/Middleware/Spa/SpaProxy/src/SpaProxyMiddleware.cs

Additional encode
